# Equipe "Equipe Exemplo em sala"

## Equipe (max 4 pessoas):
- Miguel - 123123
- Nome2 - Mat.

# Nome projeto
Classificar Default Credit Card.

# Tema
Pagamentos em cartão de crédito.

## Base de dados
Vamos trabalhar na base de Default Credit Card,
"Devedor ou não de cartão de crédito"
https://archive.ics.uci.edu/dataset/350/default+of+credit+card+clients

## Modelo/Alg.

## Objetivos
Treinar um modelo classificador para avaliar Default em Cartão de crédito.

## Etapas de modelagem
- Leitura da base de dados
- Breve EDA
- Ajuste dos dados
- Experimento 1
- Experimento 2
- Conclusões

## Leitura da base de dados

In [1]:
# prompt: Faça a leitura da base de dados https://archive.ics.uci.edu/dataset/350/default+of+credit+card+clients na variavel dados

import pandas as pd

# URL da base de dados
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/00350/default%20of%20credit%20card%20clients.xls"

# Lê a base de dados usando pandas, pulando a primeira linha (cabeçalho)
dados = pd.read_excel(url, header=1)


## EDA


In [2]:
dados

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  \
0          1      20000    2          2         1   24      2      2     -1   
1          2     120000    2          2         2   26     -1      2      0   
2          3      90000    2          2         2   34      0      0      0   
3          4      50000    2          2         1   37      0      0      0   
4          5      50000    1          2         1   57     -1      0     -1   
...      ...        ...  ...        ...       ...  ...    ...    ...    ...   
29995  29996     220000    1          3         1   39      0      0      0   
29996  29997     150000    1          3         2   43     -1     -1     -1   
29997  29998      30000    1          2         2   37      4      3      2   
29998  29999      80000    1          3         1   41      1     -1      0   
29999  30000      50000    1          2         1   46      0      0      0   

       PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0         -1  ...          0          0          0         0       689   
1          0  ...       3272       3455       3261         0      1000   
2          0  ...      14331      14948      15549      1518      1500   
3          0  ...      28314      28959      29547      2000      2019   
4          0  ...      20940      19146      19131      2000     36681   
...      ...  ...        ...        ...        ...       ...       ...   
29995      0  ...      88004      31237      15980      8500     20000   
29996     -1  ...       8979       5190          0      1837      3526   
29997     -1  ...      20878      20582      19357         0         0   
29998      0  ...      52774      11855      48944     85900      3409   
29999      0  ...      36535      32428      15313      2078      1800   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0             0         0         0         0                           1  
1          1000      1000         0      2000                           1  
2          1000      1000      1000      5000                           0  
3          1200      1100      1069      1000                           0  
4         10000      9000       689       679                           0  
...         ...       ...       ...       ...                         ...  
29995      5003      3047      5000      1000                           0  
29996      8998       129         0         0                           0  
29997     22000      4200      2000      3100                           1  
29998      1178      1926     52964      1804                           1  
29999      1430      1000      1000      1000                           1  

[30000 rows x 25 columns]

In [5]:
#Vamos ver qual a coluna é a minha classe (Default ou não)?
dados.columns

Index(['ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_0',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'default payment next month'],
      dtype='object')

In [7]:
dados['default payment next month']
#This research employed a binary variable, default payment (Yes = 1, No = 0)

0        1
1        1
2        0
3        0
4        0
        ..
29995    0
29996    0
29997    1
29998    1
29999    1
Name: default payment next month, Length: 30000, dtype: int64

In [3]:
dados.describe()

ID       LIMIT_BAL           SEX     EDUCATION      MARRIAGE  \
count  30000.000000    30000.000000  30000.000000  30000.000000  30000.000000   
mean   15000.500000   167484.322667      1.603733      1.853133      1.551867   
std     8660.398374   129747.661567      0.489129      0.790349      0.521970   
min        1.000000    10000.000000      1.000000      0.000000      0.000000   
25%     7500.750000    50000.000000      1.000000      1.000000      1.000000   
50%    15000.500000   140000.000000      2.000000      2.000000      2.000000   
75%    22500.250000   240000.000000      2.000000      2.000000      2.000000   
max    30000.000000  1000000.000000      2.000000      6.000000      3.000000   

                AGE         PAY_0         PAY_2         PAY_3         PAY_4  \
count  30000.000000  30000.000000  30000.000000  30000.000000  30000.000000   
mean      35.485500     -0.016700     -0.133767     -0.166200     -0.220667   
std        9.217904      1.123802      1.197186      1.196868      1.169139   
min       21.000000     -2.000000     -2.000000     -2.000000     -2.000000   
25%       28.000000     -1.000000     -1.000000     -1.000000     -1.000000   
50%       34.000000      0.000000      0.000000      0.000000      0.000000   
75%       41.000000      0.000000      0.000000      0.000000      0.000000   
max       79.000000      8.000000      8.000000      8.000000      8.000000   

       ...      BILL_AMT4      BILL_AMT5      BILL_AMT6       PAY_AMT1  \
count  ...   30000.000000   30000.000000   30000.000000   30000.000000   
mean   ...   43262.948967   40311.400967   38871.760400    5663.580500   
std    ...   64332.856134   60797.155770   59554.107537   16563.280354   
min    ... -170000.000000  -81334.000000 -339603.000000       0.000000   
25%    ...    2326.750000    1763.000000    1256.000000    1000.000000   
50%    ...   19052.000000   18104.500000   17071.000000    2100.000000   
75%    ...   54506.000000   50190.500000   49198.250000    5006.000000   
max    ...  891586.000000  927171.000000  961664.000000  873552.000000   

           PAY_AMT2      PAY_AMT3       PAY_AMT4       PAY_AMT5  \
count  3.000000e+04   30000.00000   30000.000000   30000.000000   
mean   5.921163e+03    5225.68150    4826.076867    4799.387633   
std    2.304087e+04   17606.96147   15666.159744   15278.305679   
min    0.000000e+00       0.00000       0.000000       0.000000   
25%    8.330000e+02     390.00000     296.000000     252.500000   
50%    2.009000e+03    1800.00000    1500.000000    1500.000000   
75%    5.000000e+03    4505.00000    4013.250000    4031.500000   
max    1.684259e+06  896040.00000  621000.000000  426529.000000   

            PAY_AMT6  default payment next month  
count   30000.000000                30000.000000  
mean     5215.502567                    0.221200  
std     17777.465775                    0.415062  
min         0.000000                    0.000000  
25%       117.750000                    0.000000  
50%      1500.000000                    0.000000  
75%      4000.000000                    0.000000  
max    528666.000000                    1.000000  

[8 rows x 25 columns]

In [ ]:
# um dos entendimentos na EDA é que alguns campos são categóricos. E devem ser ajustados.

In [20]:
## Balanceamento da target
dados['default payment next month'].value_counts()

default payment next month
0    23364
1     6636
Name: count, dtype: int64

In [21]:
# Observamos, por contagem dos valores que a minha base é bem desbalanceada.

# Ajuste dos dados

In [8]:
# prompt: para as colunas  'SEX', 'EDUCATION', 'MARRIAGE', faça o ajuste de valores categoricos com novas colunas 0 ou 1

# Cria novas colunas com base nos valores categóricos de 'SEX', 'EDUCATION' e 'MARRIAGE'
dados = pd.get_dummies(dados, columns=['SEX', 'EDUCATION', 'MARRIAGE'], prefix=['SEX', 'EDUCATION', 'MARRIAGE'])

In [9]:
dados.columns

Index(['ID', 'LIMIT_BAL', 'AGE', 'PAY_0', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5',
       'PAY_6', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4',
       'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3',
       'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6', 'default payment next month',
       'SEX_1', 'SEX_2', 'EDUCATION_0', 'EDUCATION_1', 'EDUCATION_2',
       'EDUCATION_3', 'EDUCATION_4', 'EDUCATION_5', 'EDUCATION_6',
       'MARRIAGE_0', 'MARRIAGE_1', 'MARRIAGE_2', 'MARRIAGE_3'],
      dtype='object')

In [17]:
dados[['SEX_1', 'SEX_2', 'EDUCATION_0', 'EDUCATION_1', 'EDUCATION_2','EDUCATION_3', 'EDUCATION_4', 'EDUCATION_5', 'EDUCATION_6','MARRIAGE_0', 'MARRIAGE_1', 'MARRIAGE_2', 'MARRIAGE_3']].head()

SEX_1  SEX_2  EDUCATION_0  EDUCATION_1  EDUCATION_2  EDUCATION_3  \
0  False   True        False        False         True        False   
1  False   True        False        False         True        False   
2  False   True        False        False         True        False   
3  False   True        False        False         True        False   
4   True  False        False        False         True        False   

   EDUCATION_4  EDUCATION_5  EDUCATION_6  MARRIAGE_0  MARRIAGE_1  MARRIAGE_2  \
0        False        False        False       False        True       False   
1        False        False        False       False       False        True   
2        False        False        False       False       False        True   
3        False        False        False       False        True       False   
4        False        False        False       False        True       False   

   MARRIAGE_3  
0       False  
1       False  
2       False  
3       False  
4       False

In [18]:
# prompt: Agora divida dados X e Y, em X remova a coluna ID, e para Y use a coluna default payment next month

import pandas as pd

# URL da base de dados
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/00350/default%20of%20credit%20card%20clients.xls"

# Lê a base de dados usando pandas, pulando a primeira linha (cabeçalho)
dados = pd.read_excel(url, header=1)

# Cria novas colunas com base nos valores categóricos de 'SEX', 'EDUCATION' e 'MARRIAGE'
dados = pd.get_dummies(dados, columns=['SEX', 'EDUCATION', 'MARRIAGE'], prefix=['SEX', 'EDUCATION', 'MARRIAGE'])

# Divide os dados em X e Y
X = dados.drop('ID', axis=1)
X = X.drop('default payment next month', axis = 1)
y = dados['default payment next month']

In [19]:
X

LIMIT_BAL  AGE  PAY_0  PAY_2  PAY_3  PAY_4  PAY_5  PAY_6  BILL_AMT1  \
0          20000   24      2      2     -1     -1     -2     -2       3913   
1         120000   26     -1      2      0      0      0      2       2682   
2          90000   34      0      0      0      0      0      0      29239   
3          50000   37      0      0      0      0      0      0      46990   
4          50000   57     -1      0     -1      0      0      0       8617   
...          ...  ...    ...    ...    ...    ...    ...    ...        ...   
29995     220000   39      0      0      0      0      0      0     188948   
29996     150000   43     -1     -1     -1     -1      0      0       1683   
29997      30000   37      4      3      2     -1      0      0       3565   
29998      80000   41      1     -1      0      0      0     -1      -1645   
29999      50000   46      0      0      0      0      0      0      47929   

       BILL_AMT2  ...  EDUCATION_1  EDUCATION_2  EDUCATION_3  EDUCATION_4  \
0           3102  ...        False         True        False        False   
1           1725  ...        False         True        False        False   
2          14027  ...        False         True        False        False   
3          48233  ...        False         True        False        False   
4           5670  ...        False         True        False        False   
...          ...  ...          ...          ...          ...          ...   
29995     192815  ...        False        False         True        False   
29996       1828  ...        False        False         True        False   
29997       3356  ...        False         True        False        False   
29998      78379  ...        False        False         True        False   
29999      48905  ...        False         True        False        False   

       EDUCATION_5  EDUCATION_6  MARRIAGE_0  MARRIAGE_1  MARRIAGE_2  \
0            False        False       False        True       False   
1            False        False       False       False        True   
2            False        False       False       False        True   
3            False        False       False        True       False   
4            False        False       False        True       False   
...            ...          ...         ...         ...         ...   
29995        False        False       False        True       False   
29996        False        False       False       False        True   
29997        False        False       False       False        True   
29998        False        False       False        True       False   
29999        False        False       False        True       False   

       MARRIAGE_3  
0           False  
1           False  
2           False  
3           False  
4           False  
...           ...  
29995       False  
29996       False  
29997       False  
29998       False  
29999       False  

[30000 rows x 33 columns]

In [22]:
## Como uma abordagem inicial para tratar desbalanceamanto (não ideal) vamos fazer undersampling.


LIMIT_BAL  AGE  PAY_0  PAY_2  PAY_3  PAY_4  PAY_5  PAY_6  BILL_AMT1  \
0          20000   24      2      2     -1     -1     -2     -2       3913   
1         120000   26     -1      2      0      0      0      2       2682   
2          90000   34      0      0      0      0      0      0      29239   
3          50000   37      0      0      0      0      0      0      46990   
4          50000   57     -1      0     -1      0      0      0       8617   
...          ...  ...    ...    ...    ...    ...    ...    ...        ...   
29995     220000   39      0      0      0      0      0      0     188948   
29996     150000   43     -1     -1     -1     -1      0      0       1683   
29997      30000   37      4      3      2     -1      0      0       3565   
29998      80000   41      1     -1      0      0      0     -1      -1645   
29999      50000   46      0      0      0      0      0      0      47929   

       BILL_AMT2  ...  EDUCATION_1  EDUCATION_2  EDUCATION_3  EDUCATION_4  \
0           3102  ...        False         True        False        False   
1           1725  ...        False         True        False        False   
2          14027  ...        False         True        False        False   
3          48233  ...        False         True        False        False   
4           5670  ...        False         True        False        False   
...          ...  ...          ...          ...          ...          ...   
29995     192815  ...        False        False         True        False   
29996       1828  ...        False        False         True        False   
29997       3356  ...        False         True        False        False   
29998      78379  ...        False        False         True        False   
29999      48905  ...        False         True        False        False   

       EDUCATION_5  EDUCATION_6  MARRIAGE_0  MARRIAGE_1  MARRIAGE_2  \
0            False        False       False        True       False   
1            False        False       False       False        True   
2            False        False       False       False        True   
3            False        False       False        True       False   
4            False        False       False        True       False   
...            ...          ...         ...         ...         ...   
29995        False        False       False        True       False   
29996        False        False       False       False        True   
29997        False        False       False       False        True   
29998        False        False       False        True       False   
29999        False        False       False        True       False   

       MARRIAGE_3  
0           False  
1           False  
2           False  
3           False  
4           False  
...           ...  
29995       False  
29996       False  
29997       False  
29998       False  
29999       False  

[30000 rows x 33 columns]

In [23]:
# prompt: faça undersampling para valores 0 em default payment next month para equilibrar minha base de dados em X e Y

from imblearn.under_sampling import RandomUnderSampler

# ... (código anterior) ...

# Inicializa o RandomUnderSampler
rus = RandomUnderSampler(random_state=42)  # Defina uma semente aleatória para reprodutibilidade

# Aplica o undersampling aos dados X e y
X_resampled, y_resampled = rus.fit_resample(X, y)

# Agora X_resampled e y_resampled contêm os dados balanceados
# Verifique a contagem dos valores na coluna target após o undersampling
print(y_resampled.value_counts())

# ... (código posterior) ...

default payment next month
0    6636
1    6636
Name: count, dtype: int64


In [26]:
y_resampled.value_counts()

default payment next month
0    6636
1    6636
Name: count, dtype: int64

In [27]:
#sobrescrever os nomes:
X, y = X_resampled, y_resampled

In [37]:
X.shape

(13272, 33)

In [31]:
X.dtypes

LIMIT_BAL      int64
AGE            int64
PAY_0          int64
PAY_2          int64
PAY_3          int64
PAY_4          int64
PAY_5          int64
PAY_6          int64
BILL_AMT1      int64
BILL_AMT2      int64
BILL_AMT3      int64
BILL_AMT4      int64
BILL_AMT5      int64
BILL_AMT6      int64
PAY_AMT1       int64
PAY_AMT2       int64
PAY_AMT3       int64
PAY_AMT4       int64
PAY_AMT5       int64
PAY_AMT6       int64
SEX_1           bool
SEX_2           bool
EDUCATION_0     bool
EDUCATION_1     bool
EDUCATION_2     bool
EDUCATION_3     bool
EDUCATION_4     bool
EDUCATION_5     bool
EDUCATION_6     bool
MARRIAGE_0      bool
MARRIAGE_1      bool
MARRIAGE_2      bool
MARRIAGE_3      bool
dtype: object

In [29]:
y

7510     0
15325    0
18666    0
7494     0
1239     0
        ..
29991    1
29994    1
29997    1
29998    1
29999    1
Name: default payment next month, Length: 13272, dtype: int64

## Treinamento

In [36]:
# prompt: faça  modelagem usando kfold cross validaton k=10

from sklearn.model_selection import KFold
from sklearn.tree import  DecisionTreeClassifier
from sklearn.metrics import accuracy_score
import numpy as np


# Inicializa o KFold com k=10
kf = KFold(n_splits=10, shuffle=True, random_state=42) # Defina uma semente aleatória para reprodutibilidade

# Inicializa a lista para armazenar as pontuações de acurácia
accuracy_scores = []

# Inicializa o modelo (exemplo: Regressão Logística)
model =  DecisionTreeClassifier(random_state=42)

# Loop pelos folds
for train_index, test_index in kf.split(X):
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]

    # Treina o modelo
    model.fit(X_train, y_train)

    # Faz previsões
    y_pred = model.predict(X_test)

    # Calcula a acurácia
    accuracy = accuracy_score(y_test, y_pred)
    accuracy_scores.append(accuracy)

# Imprime as pontuações de acurácia para cada fold
print("Acurácia para cada fold:", accuracy_scores)

# Calcula a acurácia média
mean_accuracy = np.mean(accuracy_scores)
print("Acurácia média:", mean_accuracy)

Acurácia para cada fold: [0.6197289156626506, 0.6197289156626506, 0.6382818387339865, 0.613413715146948, 0.613413715146948, 0.6111529766390355, 0.6179351921627732, 0.6254709871891485, 0.6412961567445365, 0.6269781461944235]
Acurácia média: 0.6227400559283099


Conclusões:

- Treinamos uma Árvore de Decisão
- Notamos que na base de dados temos campos que podem fazer discriminação sexista e etarista. Extremamente importante discutir o uso correto desses dados, com relação a computação justa https://fairlearn.org/
- E para os ajustes da base de dados o resultado inicial não se mostrou tão bom, cabendo novos experimentos. Acurácia média: 0.62 (pior caso é o 0.5 'moeda')
- Notamos na EDA que a target é desbalanceada. Usamos undersampling (de 23k para 6k exemplos na classe majoritária - 17k exemplos 'jogados fora'). Melhores aborgagens devem ser usadas, inclusive a métrica de performance, como Precision e Recall.